pkg/agent/openflow/pipeline.go

```diff
@@ -1971,46 +1973,57 @@ func (c *client) hairpinSNATFlow(endpointIP net.IP) binding.Flow {
 // serviceLBTable to trigger the learn flow, the learn flow will then send packets
 // to endpointDNATTable. Otherwise, buckets will resubmit packets to
 // endpointDNATTable directly.
-func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAffinity bool, endpoints ...proxy.Endpoint) binding.Group {
+func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, endpoints ...proxy.Endpoint) binding.Group {
 	group := c.bridge.CreateGroup(groupID).ResetBuckets()
-	var resubmitTableID binding.TableIDType
-	var lbResultMark uint32
-	if withSessionAffinity {
-		resubmitTableID = serviceLBTable
-		lbResultMark = marksRegServiceNeedLearn
-	} else {
-		resubmitTableID = endpointDNATTable
-		lbResultMark = marksRegServiceSelected
-	}
 
 	for _, endpoint := range endpoints {
-		endpointPort, _ := endpoint.Port()
 		endpointIP := net.ParseIP(endpoint.IP())
-		portVal := portToUint16(endpointPort)
 		ipProtocol := getIPProtocol(endpointIP)
 		if ipProtocol == binding.ProtocolIP {
 			ipVal := binary.BigEndian.Uint32(endpointIP.To4())
 			group = group.Bucket().Weight(100).
 				LoadReg(int(endpointIPReg), ipVal).
-				LoadRegRange(int(endpointPortReg), uint32(portVal), endpointPortRegRange).
-				LoadRegRange(int(serviceLearnReg), lbResultMark, serviceLearnRegRange).
-				LoadRegRange(int(marksReg), macRewriteMark, macRewriteMarkRange).
-				ResubmitToTable(resubmitTableID).
+				ResubmitToTable(endpointRegSetTable).
 				Done()
 		} else if ipProtocol == binding.ProtocolIPv6 {
 			ipVal := []byte(endpointIP)
 			group = group.Bucket().Weight(100).
 				LoadXXReg(int(endpointIPv6XXReg), ipVal).
-				LoadRegRange(int(endpointPortReg), uint32(portVal), endpointPortRegRange).
-				LoadRegRange(int(serviceLearnReg), lbResultMark, serviceLearnRegRange).
-				LoadRegRange(int(marksReg), macRewriteMark, macRewriteMarkRange).
-				ResubmitToTable(resubmitTableID).
+				ResubmitToTable(endpointRegSetTable).
 				Done()
 		}
 	}
 	return group
 }
 
+// endpointRegSetFlow generates the flow which sets endpoint port, service learn status and marks registers.
+func (c *client) endpointRegSetFlow(clusterIP net.IP, port int, withSessionAffinity bool, endpoint proxy.Endpoint) binding.Flow {
+	var resubmitTableID binding.TableIDType
+	var lbResultMark uint32
+	if withSessionAffinity {
+		resubmitTableID = serviceLBTable
+		lbResultMark = marksRegServiceNeedLearn
+	} else {
+		resubmitTableID = endpointDNATTable
+		lbResultMark = marksRegServiceSelected
+	}
+
+	endpointPort, _ := endpoint.Port()
+	endpointIP := net.ParseIP(endpoint.IP())
+	portVal := portToUint16(endpointPort)
+	ipProtocol := getIPProtocol(endpointIP)
+	return c.pipeline[endpointRegSetTable].BuildFlow(priorityNormal).
+		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+		MatchProtocol(ipProtocol).
+		MatchDstIP(clusterIP).
+		MatchDstPort(portToUint16(port), nil).
```

Comment:
I must be missing something obvious but why are we matching on the Cluster IP? With the current code, don't we end up with the same match for all the endpoints?

Reply:
> @hongliangl : I would prefer to add descriptions about the actual changes in the commit title and message. We can list the issue number in the commit message too.

Thanks for reminding.

Reply:
In this design, I thought of that when there are two Services(one withSessionAffinity is true and the other not) that has the same Endpoint and it is necessary to identify them with Cluster IP and port.  
However, after discussed with @tnqn, he gave more simple design and it can work as expected. I'll push the code later.